pkg/operator/ceph/csi/csi-spec.go

```diff
@@ -0,0 +1,249 @@
+/*
+Copyright 2019 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"fmt"
+
+	apps "k8s.io/api/apps/v1beta2"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type CSISpec struct {
+	Namespace string
+
+	RBDPluginImage    string
+	CephFSPluginImage string
+	RegistrarImage    string
+	ProvisionerImage  string
+	AttacherImage     string
+	//TODO: enable snapshot
+
+	CephFSProvisionerSA string
+	CephFSPluginSA      string
+
+	RBDAttacherSA    string
+	RBDProvisionerSA string
+	RBDPluginSA      string
+}
+
+var (
+	CSIParam CSISpec
+
+	EnableRBD    = false
+	EnableCephFS = false
+
+	// image names
+	DefaultRBDPluginImage    = "quay.io/cephcsi/rbdplugin:v1.0.0"
+	DefaultCephFSPluginImage = "quay.io/cephcsi/cephfsplugin:v1.0.0"
+	DefaultRegistrarImage    = "quay.io/k8scsi/csi-node-driver-registrar:v1.0.2"
+	DefaultProvisionerImage  = "quay.io/k8scsi/csi-provisioner:canary"
+	DefaultAttacherImage     = "quay.io/k8scsi/csi-attacher:v1.0.0"
+
+	// service accounts
+	// SAs are not created by operator, they must already exist
+	DefaultCephFSProvisionerSA = "rook-csi-cephfs-provisioner-sa"
+	DefaultCephFSPluginSA      = "rook-csi-cephfs-plugin-sa"
+
+	DefaultRBDAttacherSA    = "rook-csi-rbd-attacher-sa"
+	DefaultRBDProvisionerSA = "rook-csi-rbd-provisioner-sa"
+	DefaultRBDPluginSA      = "rook-csi-rbd-plugin-sa"
+
+	// template paths
+	RBDPluginTemplatePath      string
+	RBDProvisionerTemplatePath string
+	RBDAttacherTemplatePath    string
+
+	CephFSPluginTemplatePath      string
+	CephFSProvisionerTemplatePath string
+
+	rbdPlugin, cephfsPlugin                                 *apps.DaemonSet
+	rbdProvisioner, rbdAttacher, cephfsProvisioner          *apps.StatefulSet
+	rbdProvisionerSvc, rbdAttacherSvc, cephfsProvisionerSvc *corev1.Service
+)
+
+const (
+	KubeMinMajor = "1"
+	KubeMinMinor = "13"
+	dummyPort    = 1234
+	ExitOnError  = false // don't exit if CSI fails to deploy. Switch to true when flexdriver is disabled
+)
+
+func CSIEnabled() bool {
+	if EnableRBD || EnableCephFS {
+		return true
+	}
+	return false
+}
+
+func SetCSINamespace(namespace string) {
+	CSIParam.Namespace = namespace
+}
+
+func ValidateCSIParam() error {
```

Comment:
It seems that  `provisioner` and `registrar` are in both conditions so perhaps we check them individually before checking for anything `rbd` or `cephfs` specific?